File: dart/ductus/lib/src/adapter/graph_model.dart
/// Internal graph model of the Dart adapter + canonical serialization
/// into the journey graph JSON.
library;

import 'dart:collection';
import 'dart:convert';

/// Must match `version:` in pubspec.yaml — guarded by a regression test
/// in test/cli_integration_test.dart.
const String adapterVersion = '0.3.0';

const String schemaVersion = '1.0';

/// meta.adapters name of the adapter CLI (scan via `dart run ductus:adapter`).
const String cliAdapterName = 'dart';

/// meta.adapters name of the build_runner builder (path D).
const String builderAdapterName = 'dart-builder';

/// Artifact of the build_runner builder in the target package's project root
/// (path D). NOT to be confused with `ductus_graph.g.json` — that is the
/// debug file the adapter CLI writes on every scan.
const String builderArtifactFileName = 'ductus_builder.g.json';

/// Schema major version supported by the adapter (validation rule V6,
/// same as in the core).
const int supportedSchemaMajor = 1;

/// V6 logic as in the core: "major.minor" with a supported major ⇒ compatible
/// (minor extensions of the schema are backwards compatible).
bool isSupportedSchemaVersion(String version) {
  final match = RegExp(r'^(\d+)\.(\d+)$').firstMatch(version);
  return match != null && int.parse(match.group(1)!) == supportedSchemaMajor;
}

/// Origin of a graph element: manually annotated or derived.
class SourceKind {
  static const String annotation = 'annotation';
  static const String derived = 'derived';
}

/// Valid trigger values of a transition (matches `JourneyTrigger`).
const Set<String> validTriggers = {
  'tap',
  'submit',
  'auto',
  'back',
  'deeplink',
  'system',
};

/// Back-reference into the source code. [file] is always project-relative
/// with '/' separators.
class SourceRef {
  final String file;
  final int line;
  final String? symbol;

  const SourceRef({required this.file, required this.line, this.symbol});

  Map<String, Object?> toJson() => {
        'file': file,
        'line': line,
        if (symbol != null) 'symbol': symbol,
      };

  @override
  String toString() => '$file:$line';
}

/// Screen or decision node (the Dart adapter emits no action nodes;
/// actions are mapped directly to edges).
class GraphNode {
  final String id;
  final String type; // 'screen' | 'decision'
  final String? title;
  final String? flow;
  final String? description;
  final List<String> tags;
  final String source;
  final SourceRef sourceRef;

  const GraphNode({
    required this.id,
    required this.type,
    this.title,
    this.flow,
    this.description,
    this.tags = const [],
    required this.source,
    required this.sourceRef,
  });

  GraphNode copyWith({String? flow}) => GraphNode(
        id: id,
        type: type,
        title: title,
        flow: flow ?? this.flow,
        description: description,
        tags: tags,
        source: source,
        sourceRef: sourceRef,
      );

  Map<String, Object?> toJson() => {
        'id': id,
        'type': type,
        if (title != null) 'title': title,
        if (flow != null) 'flow': flow,
        if (description != null) 'description': description,
        if (tags.isNotEmpty) 'tags': [...tags]..sort(),
        'source': source,
        'sourceRef': sourceRef.toJson(),
      };
}

/// Transition (edge). [id] is optional until id generation in the merger.
class GraphEdge {
  final String? id;
  final String from;
  final String to;
  final String? trigger;
  final String? label;
  final String? condition;
  final String source;
  final SourceRef sourceRef;

  const GraphEdge({
    this.id,
    required this.from,
    required this.to,
    this.trigger,
    this.label,
    this.condition,
    required this.source,
    required this.sourceRef,
  });

  Map<String, Object?> toJson() => {
        'id': id!,
        'from': from,
        'to': to,
        if (trigger != null) 'trigger': trigger,
        if (label != null) 'label': label,
        if (condition != null) 'condition': condition,
        'source': source,
        'sourceRef': sourceRef.toJson(),
      };
}

/// Named flow. [source]/[sourceRef] are only relevant internally for merge
/// precedence and are not serialized.
class GraphFlow {
  final String id;
  final String? title;
  final String? start;
  final String? description;
  final String source;
  final SourceRef sourceRef;

  const GraphFlow({
    required this.id,
    this.title,
    this.start,
    this.description,
    required this.source,
    required this.sourceRef,
  });

  Map<String, Object?> toJson() => {
        'id': id,
        if (title != null) 'title': title,
        if (start != null) 'start': start,
        if (description != null) 'description': description,
      };
}

/// Error that terminates the adapter with a non-zero exit code;
/// [messages] go to stderr.
class AdapterException implements Exception {
  final List<String> messages;

  AdapterException(this.messages);

  @override
  String toString() => messages.join('\n');
}

Object? _canonicalize(Object? value) {
  if (value is Map) {
    final sorted = SplayTreeMap<String, Object?>();
    value.forEach((key, v) => sorted[key as String] = _canonicalize(v));
    return sorted;
  }
  if (value is List) return value.map(_canonicalize).toList();
  return value;
}

/// Canonical, diff-stable graph JSON: recursively sorted keys, 2-space
/// indent, LF, trailing newline, no `generatedAt`.
///
/// [adapterName] is the meta.adapters entry: [cliAdapterName] for the CLI
/// scan, [builderAdapterName] for the build_runner artifact (path D) —
/// otherwise both outputs are byte-identical (parity guarantee).
String encodeCanonicalGraph({
  required List<GraphFlow> flows,
  required List<GraphNode> nodes,
  required List<GraphEdge> edges,
  String adapterName = cliAdapterName,
}) {
  int byId(String a, String b) => a.compareTo(b);
  final graph = <String, Object?>{
    'schemaVersion': schemaVersion,
    'flows': ([...flows]..sort((a, b) => byId(a.id, b.id)))
        .map((f) => f.toJson())
        .toList(),
    'nodes': ([...nodes]..sort((a, b) => byId(a.id, b.id)))
        .map((n) => n.toJson())
        .toList(),
    'edges': ([...edges]..sort((a, b) => byId(a.id!, b.id!)))
        .map((e) => e.toJson())
        .toList(),
    'meta': {
      'adapters': [
        {'name': adapterName, 'version': adapterVersion},
      ],
    },
  };
  final encoded =
      const JsonEncoder.withIndent('  ').convert(_canonicalize(graph));
  return '$encoded\n';
}

File: packages/adapter-typescript/src/graph-model.ts
/**
 * Internal graph model of the TypeScript adapter + canonical serialization
 * into the journey graph JSON — semantic mirror of
 * dart/ductus/lib/src/adapter/graph_model.dart.
 */

import { createRequire } from 'node:module';
import { SCHEMA_VERSION } from '@ductus/schema';

/**
 * Adapter version from the package's own package.json — read at runtime so
 * that changesets version bumps cannot orphan a constant (the Dart adapter
 * needs a manually maintained constant plus a regression test for this;
 * here the test in test/cli.test.ts only guards the read path).
 */
export const adapterVersion: string = (
  createRequire(import.meta.url)('../package.json') as { version: string }
).version;

export const schemaVersion = SCHEMA_VERSION;

/** meta.adapters name of the adapter CLI. */
export const cliAdapterName = 'typescript';

/** Origin of a graph element: manually annotated or derived. */
export const SourceKind = {
  annotation: 'annotation',
  derived: 'derived',
} as const;

export type SourceKindValue = (typeof SourceKind)[keyof typeof SourceKind];

/** Valid trigger values of a transition (matches `TriggerType`). */
export const validTriggers = new Set(['tap', 'submit', 'auto', 'back', 'deeplink', 'system']);

/**
 * Back-reference into the source code. `file` is always project-relative
 * with '/' separators.
 */
export interface SourceRef {
  file: string;
  line: number;
  symbol?: string;
}

export function refToString(ref: SourceRef): string {
  return `${ref.file}:${ref.line}`;
}

/**
 * Screen or decision node (the adapter emits no action nodes; actions are
 * mapped directly to edges).
 */
export interface GraphNode {
  id: string;
  type: 'screen' | 'decision';
  title?: string;
  flow?: string;
  description?: string;
  tags: string[];
  source: SourceKindValue;
  sourceRef: SourceRef;
}

/** Transition (edge). `id` is optional until id generation in the merger. */
export interface GraphEdge {
  id?: string;
  from: string;
  to: string;
  trigger?: string;
  label?: string;
  condition?: string;
  source: SourceKindValue;
  sourceRef: SourceRef;
}

/**
 * Named flow. `source`/`sourceRef` are only relevant internally for merge
 * precedence and are not serialized.
 */
export interface GraphFlow {
  id: string;
  title?: string;
  start?: string;
  description?: string;
  source: SourceKindValue;
  sourceRef: SourceRef;
}

/** Error that terminates the adapter with exit ≠0; `messages` go to stderr. */
export class AdapterException extends Error {
  readonly messages: string[];

  constructor(messages: string[]) {
    super(messages.join('\n'));
    this.name = 'AdapterException';
    this.messages = messages;
  }
}

function nodeToJson(node: GraphNode): Record<string, unknown> {
  return {
    id: node.id,
    type: node.type,
    ...(node.title !== undefined ? { title: node.title } : {}),
    ...(node.flow !== undefined ? { flow: node.flow } : {}),
    ...(node.description !== undefined ? { description: node.description } : {}),
    ...(node.tags.length > 0 ? { tags: [...node.tags].sort() } : {}),
    source: node.source,
    sourceRef: refToJson(node.sourceRef),
  };
}

function edgeToJson(edge: GraphEdge): Record<string, unknown> {
  return {
    id: edge.id!,
    from: edge.from,
    to: edge.to,
    ...(edge.trigger !== undefined ? { trigger: edge.trigger } : {}),
    ...(edge.label !== undefined ? { label: edge.label } : {}),
    ...(edge.condition !== undefined ? { condition: edge.condition } : {}),
    source: edge.source,
    sourceRef: refToJson(edge.sourceRef),
  };
}

function flowToJson(flow: GraphFlow): Record<string, unknown> {
  return {
    id: flow.id,
    ...(flow.title !== undefined ? { title: flow.title } : {}),
    ...(flow.start !== undefined ? { start: flow.start } : {}),
    ...(flow.description !== undefined ? { description: flow.description } : {}),
  };
}

function refToJson(ref: SourceRef): Record<string, unknown> {
  return {
    file: ref.file,
    line: ref.line,
    ...(ref.symbol !== undefined ? { symbol: ref.symbol } : {}),
  };
}

function canonicalize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonicalize);
  if (value !== null && typeof value === 'object') {
    const sorted: Record<string, unknown> = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      sorted[key] = canonicalize((value as Record<string, unknown>)[key]);
    }
    return sorted;
  }
  return value;
}

/**
 * Canonical, diff-stable graph JSON: recursively sorted keys, 2-space
 * indent, LF, trailing newline, no `generatedAt` — byte-stable across
 * repeated runs (NFR2).
 */
export function encodeCanonicalGraph(input: {
  flows: GraphFlow[];
  nodes: GraphNode[];
  edges: GraphEdge[];
}): string {
  const byId = (a: string, b: string): number => (a < b ? -1 : a > b ? 1 : 0);
  const graph = {
    schemaVersion,
    flows: [...input.flows].sort((a, b) => byId(a.id, b.id)).map(flowToJson),
    nodes: [...input.nodes].sort((a, b) => byId(a.id, b.id)).map(nodeToJson),
    edges: [...input.edges].sort((a, b) => byId(a.id!, b.id!)).map(edgeToJson),
    meta: {
      adapters: [{ name: cliAdapterName, version: adapterVersion }],
    },
  };
  return `${JSON.stringify(canonicalize(graph), null, 2)}\n`;
}
